packages/dds/tree/src/feature-libraries/sequence-field/rebase.ts

```diff
@@ -48,57 +49,60 @@ export type NodeChangeRebaser<TNodeChange> = (
-            // We ignore #1 because slices are not yet supported.
-            // We ignore #2 because we do not yet support specifying the tiebreak.
-            factory.pushContent(clone(currMark));
-            baseIter.push(baseMark);
-        } else if (isReattach(currMark)) {
-            // We currently ignore the ways in which base marks could affect re-attaches.
-            // These are:
-            // 1. A reattach that targets the same tombs.
-            // 2. Attaches that target the same gap.
-            // We ignore #1 because it could only occur if undo were supported.
-            // We ignore #2 because we do not yet support specifying the tiebreak.
-            factory.pushContent(clone(currMark));
-            baseIter.push(baseMark);
-        } else if (isReattach(baseMark)) {
-            // We currently ignore the ways in which curr marks overlap with this re-attach.
-            // These are:
-            // 1. A reattach that matches this re-attach.
-            // 2. A tomb that matches this re-attach.
-            // We ignore #1 because it could only occur if undo were supported.
-            // We ignore #2 because we do not yet produce tombs.
-            factory.pushOffset(getOutputLength(baseMark));
-            currIter.push(currMark);
+        if (baseMark === undefined) {
+            assert(
+                currMark !== undefined,
+                "Loop condition should prevent both iterators from being empty",
+            );
+            if (baseDetachOffset > 0 && isAttach(currMark)) {
+                currIter.pop();
+                handleCurrAttach(currMark, factory, lineageRequests, baseDetachOffset);
+            } else {
+                break;
+            }
+        } else if (currMark === undefined) {
+            assert(
+                baseMark !== undefined,
+                "Loop condition should prevent both iterators from being empty",
+            );
+            baseIter.pop();
+            if (isRelevantToLineage(baseMark)) {
+                baseDetachOffset += getInputLength(baseMark);
+            } else if (!isAttach(baseMark)) {
+                break;
+            }
+        } else if (isAttach(currMark)) {
+            if (isAttach(baseMark) && isAttachAfterBaseAttach(currMark, baseMark)) {
+                baseIter.pop();
+                factory.pushOffset(getOutputLength(baseMark));
+            } else {
+                // TODO: Handle tiebreaking is baseMark is also an attach.
+                currIter.pop();
+                handleCurrAttach(currMark, factory, lineageRequests, baseDetachOffset);
+            }
         } else if (isAttach(baseMark)) {
-            // We currently ignore the ways in which curr marks overlap with these attaches.
-            // These are:
-            // 1. Slice ranges that include prior insertions
-            // We ignore #1 because we do not yet support slices.
+            baseIter.pop();
```

Comment:
Do we not need to do this stuff here:
```
if (baseDetachOffset > 0 && baseRevision !== undefined) {
    updateLineage(lineageRequests, baseRevision);
}

lineageRequests.length = 0;
baseDetachOffset = 0;
```

Reply:
I don't see why it would be necessary. In fact, it would be incorrect to do that here. In the scheme implemented in this PR, we can only record one lineage event per base changeset. A lineage event currently represents a detach of a contiguous range of nodes. The base changeset could have a detach, immediately followed by an attach, immediately followed by another detach. In that case a contiguous range of nodes is being detached by the two detaches, and so we must record it as a single lineage event.

Reply:
Ah, I understand better now. It's probably painfully obvious at that point but I find the logic really hard to follow (and therefore hard to maintain in the future). I don't have a suggestion for making it clearer but I'll think about it more.

Reply:
I'd rather not spend too much time on this for this PR. We will have to make changes to the lineage representation to support slice moves and also to enable marks to share a lineage graph, and I suspect the algorithm used here will also have to be changed to accommodate.

Reply:
Those future changes may not happen for months, and when they do, I'd rather the code being changed be simpler so we can more easily think about the changes. That said, if you're satisfied with it as is and I don't have something better to suggest, then I wouldn't want hold the PR up either. Resolving this thread. I'll pipe up if I do have something to suggest.